content/post/2020-01-02-cran-error/index.Rmd

```diff
@@ -0,0 +1,101 @@
+
+## OMG, CRAN wrote to me!
+
+There will come the time when you check your mailbox and see this dreaded message about a check that your beloved CRAN package failed. Even though the deadline for submitting a fix is quite reasonable, you might imagine yourself sitting long into the night over your laptop just to find out where is this one comma that was the source of the error. However, I need to say we were positively surprised over the level of details of feedback we got from CRAN concerning the error in _our_ package.
+
+Thus, important :sparkles: tip \#1 :sparkles: *read the e-mail from CRAN carefully!*
+
+## Reproducing the error
+
+OK, let's get to work! Firstly, check the "checks" webpage on CRAN. These CRAN checks come in different _flavors_ and hopefully, not all checks failed. In our case, the flavor that produced the error was `r-devel-linux-x86_64-debian-clang` and I had no wish to install the developmental version of R and other packages on my local laptop just to try mimicking this setup. That's when I thought about checking [*docker*](https://www.docker.com/).
+
+Installing `docker` and running a test went well, but what one needs is a VM that would be exactly the same as the platforms CRAN uses for testing. I realized that it would take too much time to re-create it, so again my DuckDuckGo search engine[^1] came to the rescue.
+
+:sparkles: Tip \#2 :sparkles: *search the net!*
+
+## R and docker
+
+<!-- r-hub and rocker - which is good for what -->
+There are two services that are worth mentioning:
+
+- [rhub](https://r-hub.github.io/rhub/), which is an R package for using the r-hub,
+- and the [Rocker project](https://www.rocker-project.org/), which distributes various docker images useful for R and RStudio testing
+
+If you are a maintainer of an R package and don't want to install docker locally, you will find the `rhub` very useful, as you can push your package to the server for testing with one command! Check out e.g., [this blog post](https://blog.r-hub.io/2019/04/25/r-devel-linux-x86-64-debian-clang/)
+
+If you want to test more locally, and importantly, if you want to test visualizations, GUI, or specific behavior of the code in RStudio, you should take a look on the `Rocker project`. They provide easy-to-run docker images with pre-installed R and RStudio, which you can play with through a browser window.
+
+## R-hub locally
+
+In my case, I am not registered as the main maintainer of Haplin, and I wanted to test locally, so I decided to use one of the `rhub` functions, [`local_check_linux()`](https://r-hub.github.io/rhub/reference/local_check_linux.html). This seemed like a very easy way to run locally tests on a platform that is exactly the same as CRAN uses! And it is... only not in my case ;)
```

Comment:
even if you're not the maintainer of a package you can use R-hub package builder using the email argument cf https://docs.r-hub.io/#email-address

Reply:
OK, I was testing this at first, but I somehow got errors when running the "check" functions and just focused on the "local_check" functions.